### ISSUE TITLE
AS3 ERROR declaration.schemaVersion must be one of the following with Controller version 2.1.0

### ISSUE THREAD

**Issue by tomaszmzyk** [F5Networks/k8s-bigip-ctlr#1467]
#### Setup Details
 CIS Version : 2.1
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP 15.1.4 
 AS3 Version: 3.21      
 Agent Mode: AS3       
 Orchestration: OSCP      
 Orchestration Version:  3.11     
 Pool Mode: Cluster      
 
#### Description  
 Hi
During a update to k8s-bigip-ctrl version 2.1.0 I see in container logs:
2020/08/18 06:23:15 [ERROR] [AS3] Template is not valid. see errors
2020/08/18 06:23:15 [ERROR] - declaration.schemaVersion: declaration.schemaVersion must be one of the following: "3.20.0", "3.19.0", "3.18.0", "3.17.0", "3.16.0", "3.15.0", "3.14.0", "3.13.0", "3.12.0", "3.11.0", "3.10.0", "3.9.0", "3.8.0", "3.7.0", "3.6.0", "3.5.0", "3.4.0", "3.3.0", "3.2.0", "3.1.0", "3.0.0"

This errors are shown whne I don't have any configmap for F5. When I deploy a configuration for F5, controller did't synchronize data with F5, and on F5 I don't have a configuration.
ConfigMap works on version 1.14

Logs from controller:
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,715 __main__ DEBUG] config handler woken for reset
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,716 __main__ DEBUG] loaded configuration file successfully
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,716 __main__ DEBUG] NET Config: {"userFdbTunnels": [{"name": "VXLAN_TUNNEL_OCP", "records": [{"name": "0c:0c:0c:66:12:0f", "endpoint": "192.168.10.15"}, {"name": "0c:0c:0c:66:12:10", "endpoint": "192.168.10.16"}]}]}
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,716 f5_cccl.service.manager DEBUG] apply_net_config start
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,716 f5_cccl.service.validation DEBUG] Validating desired config against CCCL API schema.
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,716 f5_cccl.service.validation DEBUG] validate start
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,717 f5_cccl.service.validation DEBUG] validate took 0.00092 seconds.
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,718 root DEBUG] get WITH uri: https://10.115.96.11:443/mgmt/tm/auth/partition/OPENSHIFT AND suffix:  AND kwargs: {}
2020/08/18 06:23:11 [DEBUG] [2020-08-18 06:23:11,719 urllib3.connectionpool DEBUG] Resetting dropped connection: 10.115.96.11
2020/08/18 06:23:11 [DEBUG] [CORE] Updated 0 of 0 virtual server configs, deleted 0
2020/08/18 06:23:11 [DEBUG] [CORE] Finished syncing virtual servers  project1 (203.159147ms)
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,022 urllib3.connectionpool DEBUG] https://192.168.1.1:443 "GET /mgmt/tm/auth/partition/OPENSHIFT HTTP/1.1" 200 188
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,023 f5_cccl.bigip DEBUG] Refreshing the BIG-IP net cached state...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,024 root DEBUG] get WITH uri: https://192.168.1.1:443/mgmt/tm/auth/partition/OPENSHIFT AND suffix:  AND kwargs: {}
2020/08/18 06:23:12 [DEBUG] [CORE] Updated 0 of 0 virtual server configs, deleted 0
2020/08/18 06:23:12 [DEBUG] [CORE] Finished syncing virtual servers  openshift-monitoring (213.106158ms)
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,272 urllib3.connectionpool DEBUG] https://192.168.1.1:443 "GET /mgmt/tm/auth/partition/OPENSHIFT HTTP/1.1" 200 188
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,273 f5_cccl.bigip DEBUG] Retrieving arps from BIG-IP /OPENSHIFT...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,273 root DEBUG] get WITH uri: https://192.168.1.1:443/mgmt/tm/net/arp/ AND suffix:  AND kwargs: {'params': '$filter=partition+eq+OPENSHIFT'}
2020/08/18 06:23:12 [DEBUG] [CORE] Updated 1 of 0 virtual server configs, deleted 0
2020/08/18 06:23:12 [DEBUG] [CORE] Finished syncing virtual servers  project2 (386.931783ms)
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,541 urllib3.connectionpool DEBUG] https://192.168.1.1:443 "GET /mgmt/tm/net/arp/?$filter=partition+eq+OPENSHIFT HTTP/1.1" 200 138
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,542 f5_cccl.bigip DEBUG] Retrieving fdb tunnels from BIG-IP /OPENSHIFT...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,543 root DEBUG] get WITH uri: https://192.168.1.1:443/mgmt/tm/net/fdb/tunnel/ AND suffix:  AND kwargs: {'params': {'ver': '11.5.0'}}
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,649 urllib3.connectionpool DEBUG] https://192.168.1.1:443 "GET /mgmt/tm/net/fdb/tunnel/?ver=11.5.0 HTTP/1.1" 200 1541
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.bigip DEBUG] BIG-IP net refresh took 0.62712 seconds.
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Getting arp tasks...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Getting tunnel tasks...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Getting pre-existing tunnel update tasks...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Building task lists...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Service task queue length: 0
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Creating 0 resources...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Updating 0 resources...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] Deleting 0 resources...
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 f5_cccl.service.manager DEBUG] apply_net_config took 0.93503 seconds.
2020/08/18 06:23:12 [DEBUG] [2020-08-18 06:23:12,651 __main__ DEBUG] updating tasks finished, took 0.9358420372009277 seconds

Controller configuration:
```
         command: ["/app/bin/k8s-bigip-ctlr"]
          args: [
            "--bigip-username=$(BIGIP_USERNAME)",
            "--bigip-password=$(BIGIP_PASSWORD)",
            # Replace with the IP address or hostname of your BIG-IP device
            "--bigip-url=192.168.1.1",
            # Replace with the name of the BIG-IP partition you want to manage
            "--bigip-partition=OPENSHIFT",
            "--pool-member-type=cluster",
            # Replace tunnel for OpenShift vxlan
            "--openshift-sdn-name=/OPENSHIFT/VXLAN_TUNNEL_OCP",
            # Set manage-configmaps TRUE to manage config maps
            #"--manage-configmaps=true",
            # Login level
            "--log-level=DEBUG",
            "--log-as3-response=true",
            # Self-signed cert
            "--insecure=true"
            ]
```
 
When I enable option:
"--manage-configmaps=true"
There is no change in logs.

When I deploy application, in logs I don't see any entries regarding processing, errors, but on F5 there is no configuration deploy.

I used AS3 version 3.19 (LTS). I changed to 3.21 but still the same behavior.

 

#### Steps To Reproduce
1) Install controller in version 2.1.0
2) Check logs for controller

#### Expected Result 
Properly deployed configuration on F5 from Openshift

#### Actual Result  
No configuration from Openshift on F5

#### Observations (if any)

Regards

**Comment by cisbotctlr** [comment]
cisbot will assign the issue to one of the devs. 
 @devs, use /jira for internal tracking.

**Comment by veeramalla-f5** [comment]
@tomaszmzyk  This is a known issue with 2.1.0 and will be fixed as part of 2.1.1(releasing soon).

Solution:
1. Use a patch build we created for customers who cant wait for 2.1.1
   f5networksdevel/k8s-bigip-ctlr:as3_21Support

or 

2. Disable AS3 Validation using --validate=false

Regards,
Abhishek Veeramalla

**Comment by veeramalla-f5** [comment]
@tomaszmzyk Please let me know if it worked for you.

**Comment by tomaszmzyk** [comment]
Hi
After update to version from your comment, indeed  works without errors.
I have one more question.
Is exist documentation for all options when we deploy k8s-bigip-ctrl, for example options:

--validate=false

Thanks in advance.
Regards

**Comment by veeramalla-f5** [comment]
@tomaszmzyk , We are currently working on updating our documentation for Controller Deployment and Global Parameters. It will also be available part of our next release (2.1.1)
